### PR TITLE
interfaces: add maliit input method interface

### PR DESCRIPTION
**Pull request by Elleo**
Add an interface that allows apps to talk to a confined maliit snap (such as the Ubuntu Keyboard) for on-screen keyboard support.

This involves a brand new app armor profile, as this wasn't previously confined under click.

### CODE REVIEW

File: interfaces/builtin/maliit_input_method.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+# allow binding to the maliit dbus server for address negotiation
+dbus (bind)
+    bus=session
+    name="org.maliit.server",
+
+/usr/share/glib-2.0/schemas/ r,
+
+# maliit uses peer-to-peer dbus over a unix socket after address negotiation
+unix (bind, send, receive, listen, accept) addr="@/tmp/maliit-server/dbus-*",
```
Comment (zyga) [inline interfaces/builtin/maliit_input_method.go]: This is somewhat confusing but I assume that `@` denotes an abstract socket that has no relation to the actual filesystem (and is just a fancy string with slashes). If it is not the case bare in mind that /tmp is private per-snap.

Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: Yep, that's an abstract socket; dbus automatically converts all tmpdir sockets into abstract sockets, so it ends up as an abstract socket named with a tmpdir style path name

File: interfaces/builtin/maliit_input_method.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+#include <abstractions/dbus-session>
+
+# Find the maliit input method socket
+dbus (send, receive)
+    bus=session
+    interface="org.maliit.Server.Address"
+    path=/org/maliit/server/address
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+unix (send, receive, connect) addr=none peer=(label=###SLOT_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Is this 'type=stream'?

Also, I see 3 different IPC mechanisms:
1. org.maliit.server DBus bind in maliitInputMethodPermanentSlotAppArmor
2. org.maliit.Server.Address DBus interface in maliitInputMethodConnectedPlugAppArmor
3. @/tmp/maliit-server/dbus-* abstract socket bind/etc in maliitInputMethodPermanentSlotAppArmor
4. addr=none anonymous socket in maliitInputMethodConnectedPlugAppArmor

It seems that '1' and '2' go together, but the capitalization is different (ie, org.maliit.server in 'name' and 'path' in maliitInputMethodConnectedPlugAppArmor, but org.maliit.Server in 'interface'. If these go together, can you add a comment to maliitInputMethodConnectedSlotAppArmor and maliitInputMethodConnectedPlugAppArmor that says something about 'address negotiation' so it is understood that it goes back to the well-known DBus name you are binding to in maliitInputMethodConnectedSlotAppArmor?

'3' and '4' are different. There is this comment: "# maliit uses peer-to-peer dbus over a unix socket after address negotiation" but can you expand on that? Eg, "# maliit uses peer-to-peer dbus over a unix socket after address negotiation to do x, y and z". Furthermore, I expected to see a unix rule of this form in maliitInputMethodConnectedPlugAppArmor:
```
unix (send, receive, connect) type=stream peer=(addr=@/tmp/maliit-server/dbus-*, label=###SLOT_SECURITY_TAGS###),
```

Can you add a similar comment for the anonymous socket in '4'? Furthermore, I expected to see a unix rule of this form in maliitInputMethodConnectedSlotAppArmor:
```
unix (accept, receive, send) addr=none peer=(label=###PLUG_SECURITY_TAGS###),
```

Put another way, for each IPC mechanism, we should have a rule in ConnectedSlot to allow plugging clients to connect, a corresponding rule in ConnectedPlug to allow connecting to slotting server, and comments in both describing what goes where and why.

File: interfaces/builtin/maliit_input_method.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+var maliitInputMethodPermanentSlotSecComp = []byte(`
+connect
+recvmsg
+send
+sendto
+sendmsg
+socket
+listen
+accept4
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: You can drop connect and socket, they are in the default seccomp policy.

Please add the following:
```
accept
recvfrom
```
(and group like calls together (eg, alphabetize)).

File: interfaces/builtin/maliit_input_method.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+`)
+
+var maliitInputMethodConnectedPlugSecComp = []byte(`
+connect
+recvmsg
+send
+sendto
+sendmsg
+socket
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Please drop connect and socket and add recvfrom.

File: interfaces/builtin/basedeclaration.go [hunk 5]
```diff
@@ -345,6 +345,10 @@ slots:
       slot-snap-type:
         - core
     deny-auto-connection: true
+  maliit-input-method:
+    allow-installation:
+      slot-snap-type:
+        - app
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: Please adjust this to be:
```
maliit-input-method:
    allow-installation:
      slot-snap-type:
        - app
    deny-connection: true
```

This makes it so that a snap declaration is required to connect to the slot implementation.

The above also allows plugging snaps to auto-connect which is something we only allow for slot implementations that ensure that connecting clients are properly isolated. Is this something that an unpatched maliit server ensures? Eg, does maliit protect against key sniffing, injection and the like? If not, please change the base declaration to:
```
maliit-input-method:
    allow-installation:
      slot-snap-type:
        - app
    deny-connection: true
    deny-auto-connection: true
```

Also if not, can you describe how the Canonical slot implementation will ensure this? If it is sufficiently hardened, then we can use this restrictive base declaration but issue a snap declaration that allows auto-connection to the Canonical snap.

File: interfaces/builtin/maliit_input_method.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+# DBus accesses
+#include <abstractions/dbus-session>
+
+# allow binding to the maliit dbus server for address negotiation
+dbus (bind)
+    bus=session
+    name="org.maliit.server",
+
+/usr/share/glib-2.0/schemas/ r,
```
Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: Maliit itself doesn't, but the ubuntu-keyboard does

File: interfaces/builtin/maliit_input_method.go (hunk 2 above)
Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: There's only actually two IPC mechanisms, 1 was unnecessary and has now been removed, 4 was the other side of the /tmp/maliit-server/dbus-* unix socket, app armor identifies this as having an addr=none but a peer_addr of /tmp/maliit-server/dbus-*, so I've updated the rule to include the peer_addr and have added further comments to clarify the remaining two IPC systems (the dbus address server and the unix socket).

File: interfaces/builtin/basedeclaration.go (hunk 5 above)
Comment (Elleo) [inline interfaces/builtin/basedeclaration.go]: I've set this to deny-connection: true and deny-auto-connection: true as only the Canonical patched version of maliit completely prevents sniffing and injection by checking with unity8 to determine whether the currently communicating app has focus.

Comment (Elleo) [inline interfaces/builtin/basedeclaration.go]: I've set this to deny-connection: true and deny-auto-connection: true as only the Canonical patched version of maliit completely prevents sniffing and injection by checking with unity8 to determine whether the currently communicating app has focus.

File: interfaces/builtin/basedeclaration.go [hunk 7]
```diff
@@ -345,6 +345,12 @@ slots:
...
+    deny-connection: true
+    deny-auto-connection: true
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: +1 on the base declaration.

File: interfaces/builtin/maliit_input_method.go (hunk 6 above)
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Note, unity7 allows this by default now and the upcoming unity8 interface should too. This denial is usually harmless. Will ubuntu-keyboard `plugs: [ unity8 ]`?

File: interfaces/builtin/maliit_input_method.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: It is 2017.

File: interfaces/builtin/maliit_input_method.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+var maliitInputMethodPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as a maliit server.
+
+# DBus accesses
+#include <abstractions/dbus-session>
+
+/usr/share/glib-2.0/schemas/ r,
+
+# maliit uses peer-to-peer dbus over a unix socket after address negotiation
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: You are missing a '.' at the end of this statement.

File: interfaces/builtin/maliit_input_method.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+# which assigns an individual unix socket for all further communication
+# to happen over.
+dbus (send, receive)
+    bus=session
+    interface="org.maliit.Server.Address"
+    path=/org/maliit/server/address
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# provide access to the peer-to-peer dbus socket assigned by the address service
+unix (send, receive, connect) type=stream addr=none peer=(label=###SLOT_SECURITY_TAGS###, addr="@/tmp/maliit-server/dbus-*"),
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: In thinking about this more, because this is a glob, it allows clients to connect to other client's abstract sockets. Can this be made snap specific? Eg:
```
unix (send, receive, connect) type=stream addr=none peer=(label=###SLOT_SECURITY_TAGS###, addr="@/tmp/maliit-server/dbus-${SNAP_NAME}-*"),
```

If not, what protections are in place when a malicious client tries to use an addressed not assigned to it? Please update the preceding comment to discuss this.

File: interfaces/builtin/maliit_input_method.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+# Allow applications to communicate with the maliit address service
+# which assigns an individual unix socket for all further communication
+# to happen over.
+dbus (send, receive)
+    bus=session
+    interface="org.maliit.Server.Address"
+    path=/org/maliit/server/address
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# provide access to the peer-to-peer dbus socket assigned by the address service
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: snapd's policy style dictates that you capitalize 'Provide'.

File: interfaces/builtin/maliit_input_method.go (hunk 6 above)
Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: I not sure that it necessarily will always plug unity8, that really depends on what the kiosk case ends up looking like, since it might need to operate without unity8 then? But we could remove this from the interface and have the keyboard plug unity8 for now then figure that out later?

File: interfaces/builtin/maliit_input_method.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+`)
+
+var maliitInputMethodConnectedSlotAppArmor = []byte(`
+# Provides the maliit address service which assigns an individual unix socket
+# to each application
+dbus (send, receive)
+    bus=session
+    interface="org.maliit.Server.Address"
+    path=/org/maliit/server/address
+    peer=(label=###PLUG_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: You said in https://github.com/snapcore/snapd/pull/2793#discussion_r100557176 that the 'bind' rule was not needed, but how can this service be listening if there is no bind to a well-known name? Is the non-well-known connection name (eg, :1.xxx) given out via @/tmp/maliit-server/dbus-*? If so, can you make the comments clearer on this point? That seems backwards though-- doesn't the client talk to the a well-known address service on dbus, then that service gives out the unix socket to use? If so, why isn't the dbus bind needed (and please make the comment clearer on this-- perhaps with a step be step outline of the communications).

Furthermore, should this only use 'receive' instead of 'send, receive'? (remember the response is always allowed)

File: interfaces/builtin/maliit_input_method.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+# provide access to the peer-to-peer dbus socket assigned by the address service
+unix (receive, send) type=stream addr="@/tmp/maliit-server/dbus-*" peer=(label=###PLUG_SECURITY_TAGS###),
+`)
+
+var maliitInputMethodConnectedPlugAppArmor = []byte(`
+# Description: Allow applications to connect to a maliit socket
+# Usage: common
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Please remove 'Usage: common'.

File: interfaces/builtin/maliit_input_method.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    interface="org.maliit.Server.Address"
+    path=/org/maliit/server/address
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+# provide access to the peer-to-peer dbus socket assigned by the address service
+unix (receive, send) type=stream addr="@/tmp/maliit-server/dbus-*" peer=(label=###PLUG_SECURITY_TAGS###),
+`)
+
+var maliitInputMethodConnectedPlugAppArmor = []byte(`
+# Description: Allow applications to connect to a maliit socket
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: This is a little terse. Maybe rephrase and reference what step in the protocol this is for.

File: interfaces/builtin/maliit_input_method.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var maliitInputMethodPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as a maliit server.
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Here would be a good place to put the steps for establishing a connection (see below).

File: interfaces/builtin/maliit_input_method.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var maliitInputMethodPermanentSlotAppArmor = []byte(`
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Please use `const` instead of `var` here and for all other backend policy (PermanentSlot, ConnectedSlot, ConnectedPlug, etc). See upower_observe.go as an example.

File: interfaces/builtin/maliit_input_method.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    path=/org/maliit/server/address
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# provide access to the peer-to-peer dbus socket assigned by the address service
+unix (send, receive, connect) type=stream addr=none peer=(label=###SLOT_SECURITY_TAGS###, addr="@/tmp/maliit-server/dbus-*"),
+`)
+
+var maliitInputMethodPermanentSlotSecComp = []byte(`
+recvfrom
+recvmsg
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Please add 'recv' since you have 'send'

File: interfaces/builtin/maliit_input_method.go [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+sendto
+sendmsg
+listen
+accept
+accept4
+`)
+
+var maliitInputMethodConnectedPlugSecComp = []byte(`
+recvfrom
+recvmsg
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Please add 'recv' since you have 'send'

File: interfaces/builtin/maliit_input_method_test.go [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+		c.Assert(err, IsNil)
+		c.Assert(snippet, Not(IsNil))
+		snippet, err = s.iface.PermanentSlotSnippet(s.slot, system)
+		c.Assert(err, IsNil)
+		c.Assert(snippet, Not(IsNil))
+	}
+	snippet, err := s.iface.ConnectedSlotSnippet(s.plug, s.slot, interfaces.SecurityAppArmor)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
+}
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method_test.go]: The way TestUsedSecuritySystems is written is the old way. Please add separate tests:
- TestConnectedPlugSnippetAppArmor
- TestConnectedPlugSnippetSecComp
- TestConnectedSlotSnippetAppArmor
- TestPermanentSlotSnippetAppArmor
- TestPermanentSlotSnippetSecComp

and have each check for NotNil as well as for an expected string in the policy. See upower_observe_test.go for inspiration.

File: interfaces/builtin/maliit_input_method.go [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var maliitInputMethodPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as a maliit server.
+
+# DBus accesses
+#include <abstractions/dbus-session>
+
+/usr/share/glib-2.0/schemas/ r,
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: It's 'ok' to have here, it just doesn't have anything to do with maliit. Perhaps add a comment: `# TODO: should this be somewhere else?`

**Comment by niemeyer** [comment]
@Elleo ping

**Comment by Elleo** [comment]
@niemeyer Sorry, got distracted by a recent issue with the keyboard and the new Mir API, I'll try to get the latest changes finished up today. Unfortunately until this Mir issue is fixed I'll only be able to test them under X11 though.

File: interfaces/builtin/maliit_input_method.go (hunk 12 above)
Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: I'm not clear on why this works then (but it does), because the order of communication is that the application first connects to the "org.maliit.Server.Address" interface on the session bus to then be assigned a unix socket. I don't get any apparmor denials about this, and communication happens without any problems, so I'm a bit confused as to what's happening here.

File: interfaces/builtin/maliit_input_method.go [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+#  * An application connects to the address service: org.maliit.Server.Address.
+#  * The server responds with a private unix socket of the form
+#    @/tmp/maliit-server/dbus-* on which the server is running a peer-to-peer
+#    dbus session.
+#  * All further communication happens over this channel
+#  * An application wishing to receive input then requests that it be made the
+#    active context.
+#  * At this point maliit retrieves the application's PID based on the dbus
+#    channel and verifies with Unity 8 that the application is currently
+#    focused.
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: @tyhicks - iirc, you looked at this PID stuff wrt unity8 and focus. I can't recall the details but thought it was racy. Can you confirm?

File: interfaces/builtin/maliit_input_method.go (hunk 12 above)
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Does your testing environment have a stale (eg, unconfined) maliit running? Please strace or otherwise figure out what is happening here because the security policy looks incomplete based on the (nicely written) comment in maliitInputMethodPermanentSlotAppArmor and we need to understand what is happening.

Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: Finally spotted what was going on here, I'd mistakenly included the much more permissive dbus-session instead dbus-session-strict. That's now fixed and the bind rule is necessary to make it work (along with a couple of other rules that were missing).

File: interfaces/builtin/maliit_input_method.go [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+#  * Only if the application is focused is it then able to receive input from
+#    the on-screen keyboard.
+
+# DBus accesses
+#include <abstractions/dbus-session-strict>
+
+# allow binding to the maliit dbus server for address negotiation
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Can you rephrase this to be:
```
# Allow binding to the well-known maliit DBus service name for address negotiation
```

File: interfaces/builtin/maliit_input_method.go [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+dbus (bind)
+    bus=session
+    name="org.maliit.server",
+
+# TODO: should this be somewhere else?
+/usr/share/glib-2.0/schemas/ r,
+
+# maliit uses peer-to-peer dbus over a unix socket after address negotiation.
+# Each application has its own one-to-one communication channel with the maliit
+# server, over which all further communication happens.
```
Comment (jdstrand) [inline interfaces/builtin/maliit_input_method.go]: Can you add to the end: `Send and receive rules are in the per-snap connection policy.`

**Comment by jdstrand** [comment]
@Elleo - nice find on the dbus-session-strict and sorry I didn't notice it either. I've made a note to make sure that doesn't happen again going forward.

File: interfaces/builtin/maliit_input_method.go (hunk 21 above)
Comment (tyhicks) [inline interfaces/builtin/maliit_input_method.go]: @jdstrand @Elleo I've fought against the use of isPidFocused() for a long time now. PIDs are not a proper identifier for something like this. I have previously asked @Elleo about the use of isPidFocused() in maliit and, without investigating his response, I was appeased:

  https://code.launchpad.net/~michael-sheldon/maliit/obey-unity8-focus/+merge/298312/comments/767691

@Elleo is there any chance that we could switch maliit over to isSurfaceFocused()? AAIU, the surface id is much harder to guess and/or directly influence.

Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: @tyhicks As far as I'm aware we don't have access to the surface id, as we're attempting to link the dbus channel with an application in the server, so can only query the dbus channel for its PID. Unless we were to ask the application to supply it's own surface ID over dbus, but that'd seem like we'd be trusting the app to tell the truth about that ID?

Comment (tyhicks) [inline interfaces/builtin/maliit_input_method.go]: @elleo I think it is acceptable for the app to tell us their surface ID because Mir ultimately decides if the surface ID is focused when maliit calls isSurfaceFocused(). As of last time I looked at it, the surface ID was randomly generated and was not brute forceable in any reasonable amount of time so an untrusted client application couldn't give valid, in-focus surface IDs. The first two paragraphs in the link below are relevant:

  https://code.launchpad.net/~ken-vandine/content-hub/pasteboard/+merge/296352/comments/785060

Comment (Elleo) [inline interfaces/builtin/maliit_input_method.go]: Okay, I'll start looking into getting that implemented in maliit, would you like that done before we merge the interface?

Comment (tyhicks) [inline interfaces/builtin/maliit_input_method.go]: @Elleo I don't think it is a blocker for this interface. Can you simply add a TODO comment that this will soon change to using the surface ID and then you can treat the transition from pid to surface ID as orthogonal to this PR.

Comment (tyhicks) [inline interfaces/builtin/maliit_input_method.go]: @Elleo BTW, thanks for looking into the surface ID transition!

**Comment by niemeyer** [comment]
@Elleo Assuming everybody is happy with the above change and it goes in, and tests pass (they currently don't), I think this is ready to merge.

**Comment by jdstrand** [comment]
@niemeyer - +1 on 'maliit'. Otherwise LGTM for merging.
